### PR TITLE
[iOS] fix resource resolver image downloads

### PR DESCRIPTION
**Pull request by nesalang**
- allow resource resolver to handle async operation in the client's context.
- bring the branch to latest of the master
- added support for icon image in actions 
- added support for background image / media / media icon image / image

### CODE REVIEW

**Comment by tnev** [comment]
@nesalang Perfect! Does exactly what I was looking for. Really like how your implementation uses the delegate method that was already there so it doesn't add any more delegate methods. Thanks again for the helpfulness and responsiveness, it makes me very excited to continue using this fantastic project :)

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h [hunk 1]
```diff
@@ -17,7 +17,8 @@ typedef NS_ENUM(NSInteger, ACRActionType) {
 
 typedef NS_ENUM(NSInteger, ACRIconPlacement) {
     ACRAboveTitle = 0,
-    ACRLeftOfTitle
+    ACRLeftOfTitle,
+    ACROnTitle,
```
Comment (khouzam) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOBaseActionElement.h]: >ACROnTitle [](start = 4, length = 10)

Should this be NoTitle?

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm [hunk 2]
```diff
@@ -10,25 +10,43 @@
...
+
+    if([resolver respondsToSelector:@selector(resolveImageResource:)]) {
+        _resolversIFMap[scheme] = [NSNumber numberWithInt:ACOImageIF];
+    }
+    // only one IF per scheme is supported and ACRImageViewIF will be chosen
+    // when both are implemented
+    if([resolver respondsToSelector:@selector(resolveImageViewResource:)]) {
+        _resolversIFMap[scheme] = [NSNumber numberWithInt:ACOImageViewIF];
+    }
 }
```
Comment (khouzam) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACOResourceResolvers.mm]: This should be above and have an else.

If (supports resolveImageView)
{
    set ImageView
}
else if (supports resolveImage)
{
    set Image
}

This makes it easier to extend. First priority as the first statement

File: source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRButton.mm [hunk 3]
```diff
@@ -82,26 +83,43 @@ + (UIButton* )rootView:(ACRView *)rootView
...
+        [button addSubview:iconView];
+        button.iconView = iconView;
+        [button setImageView:img withConfig:config];
+
+    } else if(key.length) {
+        NSNumber *number = [NSNumber numberWithUnsignedLongLong:(unsigned long long)action.get()];
+        NSString *key = [number stringValue];
+        UIImageView *view = [rootView getImageView:key];
+        if(view && view.image) {
+            button.iconView = view;
```
Comment (khouzam) [inline source/ios/AdaptiveCards/AdaptiveCards/AdaptiveCards/ACRButton.mm]: What if view is null? is this possible?

File: source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m [hunk 4]
```diff
@@ -368,4 +369,11 @@ - (void)keyboardWillBeHidden:(NSNotification*)aNotification
     self.scrView.scrollIndicatorInsets = contentInsets;
 }
 
+- (void)didLoadElements
+{
+    [self.curView setNeedsLayout];
+    NSLog(@"complted loading elements");
```
Comment (khouzam) [inline source/ios/AdaptiveCards/ADCIOSVisualizer/ADCIOSVisualizer/ViewController.m]: >omplte [](start = 13, length = 6)

completed

**Pull request merged**